### PR TITLE
Use popup window for hover preview

### PR DESCRIPTION
**Pull request by mattn**
terminal7

### CODE REVIEW

File: autoload/lsp/ui/vim.vim [hunk 1]
```diff
@@ -440,9 +440,9 @@ function! s:handle_location(ctx, server, type, data) abort "ctx = {counter, list
                 let l:loc = a:ctx['list'][0]
                 let l:buffer = bufnr(l:loc['filename'])
                 if &modified && !&hidden
-                    let l:cmd = l:buffer !=# -1 ? 'sb ' . l:buffer : 'split ' . l:loc['filename']
+                    let l:cmd = l:buffer !=# -1 ? 'sb ' . l:buffer : 'split ' . fnameescape(l:loc['filename'])
                 else
-                    let l:cmd = l:buffer !=# -1 ? 'b ' . l:buffer : 'edit ' . l:loc['filename']
+                    let l:cmd = l:buffer !=# -1 ? 'b ' . l:buffer : 'edit ' . fnameescape(l:loc['filename'])
```
Comment (mattn) [inline autoload/lsp/ui/vim.vim]: Ah, sorry. had included fix about another issue.

**Comment by prabirshrestha** [comment]
Seems like there is a similar PR but for neovim at https://github.com/prabirshrestha/vim-lsp/pull/395. Would be good if we could get it working for vim8 and neovim.

1. Is it possible to focus on the popup window so that I can copy doc samples?
2. Should we add a flag so that users can decide the preference `let g:lsp_hover_ui = ['floating_window', 'preview']`
3. Allow to add custom keybindings to close the preview window.

//cc @jerdna-regeiz

**Comment by prabirshrestha** [comment]
Seems like doesn't correctly handle when the popup window is very large. This is also another scenario where I would need to focus on the popup window and scroll.

image
This is using https://github.com/theia-ide/typescript-language-server in cli.ts

**Comment by jerdna-regeiz** [comment]
I'll check it out tomorrow and see if I can easily integrate it with #395 keeping all features.

**Comment by mattn** [comment]
> Is it possible to focus on the popup window so that I can copy doc samples?

No, current implementation of popup on vim8 does not provide way to give focus.

> Should we add a flag so that users can decide the preference let g:lsp_hover_ui = ['floating_window', 'preview']

Okay, will do in later.

> Allow to add custom keybindings to close the preview window.

This too. Thank you.

**Comment by jerdna-regeiz** [comment]
I have integrated it into #395, thanks to @mattn. Feel free to join me on a combined approach!
But sadly mattn is right... there will be no focus at all. I guess the feature is pushed by other GUI development and relies on mouse to copy =(

**Comment by prabirshrestha** [comment]
Let us use the PR by @jerdna-regeiz in https://github.com/prabirshrestha/vim-lsp/pull/395 for discussions as it works for both vim8 and neovim.

**Comment by prabirshrestha** [comment]
Merged https://github.com/prabirshrestha/vim-lsp/pull/395 in favor of this PR. Thanks.

**Pull request closed without merging**